namcap.1
.TH namcap 1 "June 14, 2010" "namcap 2.6" "User Commands"
.SH NAME
namcap \- package analysis utility
.SH SYNOPSIS
\fBnamcap [options] <package|PKGBUILD> [package|PKGBUILD] ...
.SH DESCRIPTION
.PP
\fBnamcap\fP is a \fIpackage analysis\fP utility that looks for problems with Arch Linux packages or their PKGBUILD files.  It can apply rules to the file list, the files themselves, or individual PKGBUILD files.
.PP
Rules return lists of messages.  Each message can be one of three types: error, warning, or information (think of them as notes or comments).  Errors (designated by 'E:') are things that namcap is very sure are wrong and need to be fixed.  Warnings (designated by 'W:') are things that namcap thinks should be changed but if you know what you're doing then you can leave them.  Information (designated 'I:') are only shown when you use the info argument.  Information messages give information that might be helpful but isn't anything that needs changing.
.SH OPTIONS
.TP
\fB\-e\fR RULELIST, \fB\-\-exclude=\fRRULELIST
Do not run RULELIST rules on the package
.TP
.B "\-i, \-\-info"
display information messages
.TP
.B "\-m, \-\-machine\-readable"
displays easily parseable namcap tags instead of the normal human readable description; for example using non-fhs-man-page instead of "Non-FHS man page (%s) found. Use /usr/share/man instead". A full list of namcap tags along with their human readable descriptions can be found at /usr/share/namcap/tags.
.TP
\fB\-r\fR RULELIST, \fB\-\-rules=\fRRULELIST
only apply RULELIST rules to the package
.IP
RULELIST is a comma-separated list of rule names; if RULELIST=list then namcap returns a list of valid rules and their descriptions
.SH RULES
.TP
.B arrays
Verifies that array variables are actually arrays and not strings
.TP
.B badbackups
Verifies that backup entries don't begin with /
.TP
.B carch
Looks for references to i686 and i586 and tells you to use $CARCH instead
.TP
.B capsnames
Checks a PKGBUILD to verify that the package name has no upper case characters
.TP
.B capsnamespkg
Checks a package to verify that the package name has no upper case characters
.TP
.B checksums
Makes sure that a PKGBUILD includes valid checksums
.TP
.B depends
This module runs ldd on all executables, gets the link-level dependencies, finds the smallest subset of dependencies that cover the link-level dependencies, and compares that list to the depends of the package.  It returns messages in three cases: dependency detected and not included, dependency included but already satisfied, and dependency included and not needed.  These suggestions are just guidelines and all package builders should take this into account (i.e. you're smarter than namcap is)

Some cases where namcap fails are dlopen() and obscure links.  dlopen()'d libraries don't show up because they are loaded at run time: in the case of a program that loads plugins.  Obscure links are the cases where only a small portion of the package needs something to run; usually, the small portion won't be accessed unless that thing is installed (i.e. a java plugin for mozilla included in the java package).
.TP
.B directoryname
Checks the directory names in the package file list for standard directories (i.e. /etc, /usr/bin, /usr/lib).  All files in directories that are non-standard are returned
.TP
.B elffiles
Gives an error if it finds ELF (binary) files outside the standard paths (/bin, /usr/bin, etc.)
.TP
.B emptydir
Looks for directories in a package that don't contain any files
.TP
.B extravars
Verifies that extra variables start with an underscore
.TP
.B fhsinfopages
Verifies correct installation of info pages
.TP
.B fhsmanpages
Verifies correct installation of man pages
.TP
.B fileownership
Verifies that all files are owned by root:root and not some other user or group
.TP
.B gnomemime
Checks for generated GNOME mime files
.TP
.B hicoloricons
Checks whether the hicolor icon cache is updated when a package installs
files in /usr/share/icons/hicolor
.TP
.B infodirectory
Verifies that a package does not contain the info directory file (/usr/share/info/dir)
.TP
.B invalidstartdir
Any PKGBUILDs that refer to files in $startdir are incorrect
.TP
.B libtool
Warns if libtool (*.la) files are present
.TP
.B license
Verifies that the licenses variable has been filled in in a PKGBUILD.
.TP
.B licensepkg
Verifies that the licenses variable has been filled in in a package. For packages with a custom license it
checks whether the license file has been installed in
/usr/share/licenses/$pkgname/
.TP
.B mimefiles
Checks whether update-mime-database is called when the package installs files in /usr/share/mime
.TP
.B perllocal
Searches for perllocal.pod.  perllocal.pod is a nasty file that's included during most perl module installations
.TP
.B permissions
Checks basic file and and directory permissions.  It returns warnings about world readable bits not being set, world writable bits being set, and world executable bit not being set for directories
.TP
.B rpath
Gives an error if a binary has RPATH set to something other than /usr/lib
.TP
.B scrollkeeper
Verifies that there aren't any scrollkeeper directories
.TP
.B sfurl
Warns if a PKGBUILD is downloading from a specific sourceforge mirror instead of the generic downloads.sourceforge.net
.TP
.B symlink
Finds out symbolic and hard links which do not point to a valid path in the package
.TP
.B tags
For package maintainers, this module verifies that we have remembered to add a Maintainer and Contributor comment.
.TP
.B url
Verifies that we have the url variable set in the PKGBUILD
.TP
.B urlpkg
Verifies that we have the url variable set in the package file
.SH EXAMPLES
.TP
.B namcap foo.pkg.tar.gz
apply all rules to foo.pkg.tar.gz
.TP
.B namcap -r depends,symlink foo.pkg.tar.gz
apply the depends and symlink rules to foo.pkg.tar.gz
.TP
.B namcap PKGBUILD
apply all PKGBUILD based rules to the file PKGBUILD
.TP
.B namcap -r list
list all of the available rules
.SH COPYRIGHT
Copyright \(co 2003-2009 Jason Chu
.br
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
.SH AUTHORS
.nf
Jason Chu <jason@archlinux.org>
Abhishek Dasgupta <abhidg@gmail.com>
Hugo Doria <hugo@archlinux.org>
Dan McGee <dan@archlinux.org>
Allan McRae <allan@archlinux.org>
Jesse Young <jesseyoung@gmail.com>
JJDaNiMoTh <jjdanimoth@gmail.com>
.fi


NEWS
namcap (2.6)
	- Add new 'anyelf' rule; should catch ELF files that end up in 'any'
	  arch packages and cannot be arch-independent
	- Allow checking of all checksum types (FS#17168)
	- Code style and import cleanups
	- Handle optdepends descriptions correctly (FS#18259)
	- Make it an error to use $startdir in PKGBUILDs

namcap (2.5)
    - Turn namcap into a wrapper script that handles the compression;
      this allows us to use XZ-compressed packages without Python support
    - New warning- packages over 50% docs
    - Fix up rpath module a bit; ensure more paths are seen as valid
    - Improve checking for tarfiles; don't assume .pkg.tar.gz ending
    - Allow static libraries to have 644 or 444 permissions

namcap (2.4)
    - Added some vim modelines
    - Small fixes in mimefiles
    - Check for tags and version consistency
    - Rename 'tags' to 'namcap-tags'
    - Only process tags if necessary
    - Move extracted variable to the correct scope
    - Only do active_modules check once
    - Move PKGBUILD processing to a function
    - Move real package processing to a function
    - Make the depends module not suck
    - Move is_elf function to a utility file
    - Use the is_elf function in other modules where it makes sense
    - Clean up output methods

namcap (2.3)
    - New tags:
        elffile-not-in-allowed-dirs (FS#15299)
        gnome-mime-file (checks for GNOME mime files)
        mime-cache-not-updated (checks for update-mime-database calls)
        hicolor-icon-cache-not-updated (FS#15028)
        insecure-rpath (FS#14049)
        non-fhs-info-page (FHS compliance of installed info pages)
        pkgname-in-description (pkgname should not be in description)
        potential-non-fhs-info-page

    - Removed tags:
        missing-cvs-id
        maintainer-tags-for-tus-devs
        cvs-id-for-tus-devs

    - Added rule to look for .a files with wrong permissions
    - Report dangling symlinks as an error (dangling-symlink)
    - Added sf.net urls in sfurl module.
    - Support machine-parseable tags (use -m to see them)
    - Add covered packages to dependencies
    - gnomemime uses /usr/share/ directory now
    - Removed obsolete module gnomemenu
    - Namcap now prints a warning if a package is empty
    - Prevent crash when namcap fails to open a file (FS#14371)
    - Script detection generalised, supports more scripts (FS#15160)
    - Documentation updates (manual page, AUTHORS, README)
    - Add a option to exclude some rules from the checks


namcap.py
#! /usr/bin/env python
# 
# namcap - A Pacman package analyzer
# Copyright (C) 2003-2009 Jason Chu <jason@archlinux.org>
# 
#   This program is free software; you can redistribute it and/or modify
#   it under the terms of the GNU General Public License as published by
#   the Free Software Foundation; either version 2 of the License, or
#   (at your option) any later version.
#
#   This program is distributed in the hope that it will be useful,
#   but WITHOUT ANY WARRANTY; without even the implied warranty of
#   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#   GNU General Public License for more details.
#
#   You should have received a copy of the GNU General Public License
#   along with this program; if not, write to the Free Software
#   Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
#
# 

# System wide global stuff
import sys, os, os.path, imp, getopt, types, tarfile, re, string, Namcap, pacman
import shutil

sandbox_directory = '/tmp/namcap.' + str(os.getpid())

# Functions
# Return all possible modules (rules)
def get_modules():
	return Namcap.__all__

# Display usage information
def usage():
	print
	print "Usage: " + sys.argv[0] + " [OPTIONS] packages"
	print 
	print "Options are:"
	print "    -i                               : prints information responses from rules"
	print "    -m                               : makes the output parseable (machine-readable)"
	print "    -e rulelist, --exclude=rulelist  : don't apply RULELIST rules to the package"	
	print "    -r rulelist, --rules=rulelist    : only apply RULELIST rules to the package"
	
	
	sys.exit(2)

# Is the package a valid package file?
def verify_package(filename):
	if not os.path.isfile(filename):
		return 0
	if not tarfile.is_tarfile(filename):
		return 0
	try:
		tar = tarfile.open(package, "r")
		if not tar:
			return 0		
		if not len(tar.getmembers()) > 0:
			tar.close()
			return 0
		if not '.PKGINFO' in tar.getnames():
			tar.close()
			return 0
	except IOError:
		if tar:
			tar.close()
		return 0
	return tar

def process_tags(filename=None, machine=False):
	if machine:
		return lambda s: s
	tags = {}
	if not filename:
		filename="/usr/share/namcap/namcap-tags"
	f = open(filename)
	for i in f.readlines():
		if i[0] == "#" or i.strip() == "": continue
		tagdata = i[:-1].split("::")
		tags[tagdata[0].strip()] = tagdata[1].strip()

	return lambda s: tags[s]

def check_rules_exclude(optlist):	
	'''Check if the -r (--rules) and the -r (--exclude) options
	are being used at same time'''
	args_used = 0
	for i in optlist:	
		if '-r' in i or '-e' in i:
			args_used += 1
		if '--rules' in i or '--exclude' in i:
			args_used += 1			
	return args_used

def show_messages(name, key, messages):
	for msg in messages:
		print "%s %s: %s" % (name, key, m(msg[0]) % msg[1])

def process_realpackage(package, modules):
	extracted = 0
	pkgtar = verify_package(package)

	if not pkgtar:
		print "Error: " + package + " is empty or is not a valid package"
		return 1

	pkginfo = pacman.load(package)

	# Loop through each one, load them apply if possible
	for i in modules:
		cur_class = __import__('Namcap.' + i, globals(), locals(), [Namcap])
		pkg = cur_class.package()
		ret = [[], [], []]
		if pkg.type() == "tarball":
			if pkg.prereq() == "extract":
				# If it's not extracted, then extract it and then analyze the package
				if not extracted:
					os.mkdir(sandbox_directory)
					for j in pkgtar.getmembers():
						pkgtar.extract(j, sandbox_directory)
					extracted = 1
				ret = pkg.analyze(pkginfo, sandbox_directory)
			elif pkg.prereq() == "pkg":
				ret = pkg.analyze(pkginfo, None)
			elif pkg.prereq() == "tar":
				ret = pkg.analyze(pkginfo, pkgtar)
			else:
				ret = [['Error running rule (' + i + ')'], [], []]

			# Output the three types of messages
			show_messages(pkginfo.name, 'E', ret[0])
			show_messages(pkginfo.name, 'W', ret[1])
			if info_reporting:
				show_messages(pkginfo.name, 'I', ret[2])

	# Clean up if we extracted anything
	if extracted:
		shutil.rmtree(sandbox_directory)

def process_pkgbuild(package, modules):
	# We might want to do some verifying in here... but really... isn't that what pacman.load is for?
	pkginfo = pacman.load(package)

	if pkginfo == None:
		print "Error: " + package + " is not a valid PKGBUILD"
		return 1

	for i in modules:
		cur_class = __import__('Namcap.' + i, globals(), locals(), [Namcap])
		pkg = cur_class.package()
		ret = [[], [], []]
		if pkg.type() == "pkgbuild":
			ret = pkg.analyze(pkginfo, package)

		# Output the PKGBUILD messages
		name = "PKGBUILD (" + pkginfo.name + ")"
		show_messages(name, 'E', ret[0])
		show_messages(name, 'W', ret[1])
		if info_reporting:
			show_messages(name, 'I', ret[2])


# Main
modules = get_modules()
info_reporting = 0
machine_readable = False
filename = None

# get our options and process them
try:
	optlist, args = getopt.getopt(sys.argv[1:], "ihmr:e:t:",
			["info", "help", "machine-readable", "rules=", "exclude=", "tags="])
except getopt.GetoptError:
	usage()

active_modules = []

# Verifying if we are using the -r and -r options at same time
if check_rules_exclude(optlist) > 1:
	print "You cannot use '-r' (--rules) and '-e' (-exclude) options at same time"
	usage()

for i, k in optlist:
	if i in ('-r', '--rules') and active_modules == []:
		if k == 'list':
			print "-"*20 + " Namcap rule list " + "-"*20
			for j in modules:
				print string.ljust(j, 20) + ": " + __import__('Namcap.' + j, globals(), locals(), [Namcap]).package().long_name()
			sys.exit(2)

		module_list = k.split(',')
		for j in module_list:
			if j in modules:
				active_modules.append(j)
			else:
				print "Error: Rule '" + j + "' does not exist"
				usage()

	# Used to exclude some rules from the check
	if i in ('-e', '--exclude') and active_modules == []:
		module_list = k.split(',')
		active_modules = modules
		for j in module_list:
			if j in modules:
				active_modules.remove(j)
			else:
				print "Error: Rule '" + j + "' does not exist"
				usage()

	if i in ('-i', '--info'):
		info_reporting = 1

	if i in ('-h', '--help'):
		usage()
	if i in ('-m', '--machine-readable'):
		machine_readable = True

	if i in ('-t', '--tags'):
		filename = k

# If there are no args, print usage
if (args == []):
	usage()

m = process_tags(filename=filename, machine=machine_readable)
packages = args

# No rules selected?  Then select them all!
if active_modules == []:
	active_modules = modules

# Go through each package, get the info, and apply the rules
for package in packages:
	if not os.access(package, os.R_OK):
		print "Error: Problem reading " + package
		usage()

	if os.path.isfile(package) and tarfile.is_tarfile(package):
		process_realpackage(package, active_modules)
	elif package[-8:] == 'PKGBUILD':
		process_pkgbuild(package, active_modules)
	else:
		print "Error: Cannot process %s" % package

# vim: set ts=4 sw=4 noet:
